storage/umass/umass.c

```diff
@@ -242,25 +522,164 @@ static int umass_getattr(umass_dev_t *dev, int type, long long int *attr)
 	if (type != atSize)
 		return -EINVAL;
 
-	*attr = dev->partSize * UMASS_SECTOR_SIZE;
+	*attr = dev->part.sectors * UMASS_SECTOR_SIZE;
 
 	return EOK;
 }
 
 
-static umass_dev_t *umass_devFind(int id)
+static umass_dev_t *_umass_devFind(id_t id)
 {
-	umass_dev_t *tmp = umass_common.devices;
+	return lib_treeof(umass_dev_t, node, idtree_find(&umass_common.devices, id));
+}
 
-	if (tmp != NULL) {
-		do {
-			if (tmp->fileId == id)
-				return tmp;
-			tmp = tmp->next;
-		} while (tmp != umass_common.devices);
+
+static ssize_t umass_read(id_t id, off_t offs, char *buf, size_t len)
+{
+	mutexLock(umass_common.lock);
+	umass_dev_t *dev = _umass_devFind(id);
+	mutexUnlock(umass_common.lock);
+	if (dev == NULL) {
+		return -ENODEV;
 	}
+	return umass_readFromDev(dev, offs, buf, len);
+}
 
-	return NULL;
+
+static ssize_t umass_write(id_t id, off_t offs, const char *buf, size_t len)
+{
+	mutexLock(umass_common.lock);
+	umass_dev_t *dev = _umass_devFind(id);
+	mutexUnlock(umass_common.lock);
+	if (dev == NULL) {
+		return -ENODEV;
+	}
+	return umass_writeToDev(dev, offs, buf, len);
+}
+
+
+static int umass_mountFromDev(umass_dev_t *dev, const char *name, oid_t *oid)
+{
+	umass_fs_t *fs;
+	int err;
+
+	if (dev == NULL) {
+		return -ENODEV;
+	}
+
+	if (dev->part.fs != NULL) {
+		return -EEXIST;
+	}
+
+	fs = umass_getfs(name);
+	if (fs == NULL) {
+		return -ENOENT;
+	}
+	dev->part.fs = fs;
+
+	err = portCreate(&dev->part.port);
+	if (err != 0) {
+		fprintf(stderr, "umass: Can't create partition port!\n");
+		return 1;
+	}
+
+	oid->port = dev->part.port;
+	oid->id = dev->fileId;
+
+	err = fs->mount(oid, UMASS_SECTOR_SIZE, umass_read, umass_write, &dev->part.fdata);
+	if (err < 0) {
+		return err;
+	}
+	oid->id = err;
+
+	err = beginthread(umass_fsthr, 4, dev->part.fsstack, sizeof(dev->part.fsstack), &dev->part);
+	if (err < 0) {
+		dev->part.fs->unmount(dev->part.fdata);
+		dev->part.fs = NULL;
+		dev->part.fdata = NULL;
+		return err;
+	}
+
+	return EOK;
+}
+
+
+static int umass_mount(id_t id, const char *name, oid_t *oid)
+{
+	umass_dev_t *dev;
+
+	mutexLock(umass_common.lock);
+	dev = _umass_devFind(id);
+	mutexUnlock(umass_common.lock);
+
+	return umass_mountFromDev(dev, name, oid);
+}
+
+
+static void umass_fsthr(void *arg)
+{
+	umass_part_t *part = (umass_part_t *)arg;
+	umass_req_t *req;
+	int umount = 0, ret;
+
+	for (;;) {
+		req = (umass_req_t *)malloc(sizeof(umass_req_t));
```

Comment:
Not sure if using `malloc` here is the best idea, maybe leave TODO to revisit this later.